src/libraries/Microsoft.Bcl.Cryptography/src/buildTransitive/Microsoft.Bcl.Cryptography.targets

```diff
@@ -0,0 +1,8 @@
+<Project InitialTargets="NETStandardCompatError_Microsoft_Bcl_Cryptography">
+    <Target Name="NETStandardCompatError_Microsoft_Bcl_Cryptography">
+        <Error Condition="$([MSBuild]::IsTargetFrameworkCompatible('$(TargetFramework)', 'netstandard2.1')) AND
+                          !$([MSBuild]::IsTargetFrameworkCompatible('$(TargetFramework)', 'net8.0')) AND
+                          !$([MSBuild]::IsTargetFrameworkCompatible('$(TargetFramework)', 'net462'))"
+               Text="Microsoft.Bcl.Cryptography doesn't support $(TargetFramework). Use multi-targeting instead of .NET Standard 2.1 for this package, or target .NET Standard 2.0." />
```

Comment:
This file is probably the thing that needs the most attention in this pull request. It ends up as a buildTransitive target for .NET Standard 2.1. I did confirm that it does block the package installation in .NET Standard 2.1.

In particular, thoughts on wording for the message and the condition would be appreciated.

<img width="748" alt="Screenshot 2025-01-22 at 3 50 34 PM" src="https://github.com/user-attachments/assets/c2a1052f-b83c-412a-836d-bf1e6c369be6" />

Reply:
How about you move this file to `buildTransitive/netstandard2.1` and remove the conditions? [There are already](https://nuget.info/packages/Microsoft.Bcl.Cryptography/9.0.1) empty folders in `buildTransitive` for the supported frameworks, causing the file to not be imported there.

In a similar vein you can add a `_._` file in `lib/netstandard2.1`, causing no library to be imported in frameworks compatible with .NET Standard 2.1, but earlier than .NET 8. With that, the error can become a warning, which will not break scenarios when `Microsoft.Bcl.Cryptography` is referenced by an NS2.0 library, which is referenced by a NS2.1 library (the end project will still have to target .NET 8+).

Reply:
This infrastructure already exists in all shipping runtime packages. Just as a warning instead of an error for very good reasons. Can you please explain why the existing infrastructure isn't sufficient?

Reply:
The use case is different here. In API review, it was decided that we cannot outbox `AesGcm` in to M.Bcl.Cryptography for .NET Standard 2.1 because the API shape for it is a pit of failure. The API review folks basically decided "It is an error to use this package on .NET Standard 2.1". @bartonjs correct me if I'm wrong.

Reply:
It's not possible to enforce this via a package target files. Consumers can always set the `ExcludeAssets` metadata to control the assets that are being used from a package (that's a fundamental design of NuGet packages):

`<PackageReference Include="Microsoft.Bcl.Cryptography" Version="..." ExcludeAssets="buildTransitive" />`.

In general, as long as the package supports .NET Standard 2.0 or .NET Standard 2.1, it's consumable on that platform. You would need to make the assembly throw at runtime. Note that we received loud and clear feedback that throwing at runtime is the worst of all options. See https://github.com/open-telemetry/opentelemetry-dotnet/issues/3448#issuecomment-1185459615 and the following comments for more information.

Reply:
@ericstj I would appreciate your opinion on this as well

Reply:
Some other options I have thought about, in descending order of preference:

* Hide, obsolete-as-error and change the problematic APIs of `AesGcm` to unconditionally throw when targeting `netstandard2.0`.
* Multi-target to `netstandard2.1` and do not polyfill `AesGcm` there.
* Multi-target to `netcoreapp3.0` and type-forward `AesGcm` there (I can imagine you don't want this).

What do you think about my proposal above? On second thought we could add the empty folder in `lib/netcoreapp3.0` instead of `lib/netstandard2.1`.

Reply:
I agree with not trying to block things via error targets.  We add the targets to try and send a message, but not to block consumption.

If you can avoid exposing any assets on that target framework it's the best - then nuget won't allow install.  Seems that ship has sailed.

In that case, I think I like @teo-tsirpanis's suggestion the best - if we have API that is not appropriate to share in `netstandard` ref - just leave it out.   `netstandard` is lower / smaller than `net` or `netcoreapp` frameworks.

Reply:
NETStandard 2.1 defines a version of the class with two constructors that we're not exposing from this package (they've since been marked `[Obsolete]` for "no one uses these in a secure enough manner" reasons).  So the problem is that if someone runs on a NS2.1 target that isn't NET8+ they'll get a MissingMethodException, because the underlying runtime won't have exposed the ctors that this library says exist.

The choices are
1) Actively block NS2.1 (what API Review said to do)
2) Deal with NS2.1 MissingMethodException
3) Say that "because NS2.1 exists, we will not support this for .NET Framework", which probably gets someone sent to the principal's office.

Option (4) of "expose the old ctors" has already been rejected on "secure by default" grounds.

Reply:
Oh, right, not having a NS2.1 asset means they won't unify, so no MissingMethodException.  Instead it's "Ambiguous reference, System.Security.Cryptography.AesGcm [netstandard2.1] vs System.Security.Cryptography.AesGcm [Microsoft.Bcl.Cryptography]"

Reply:
Oh [there is `AesGcm` in .NET Standard 2.1](https://learn.microsoft.com/en-us/dotnet/api/system.security.cryptography.aesgcm?view=net-9.0#applies-to). Then why not multi-target to .NET Standard 2.1 and forward to the inbox `AesGcm`? Sounds like a no-brainer.

Reply:
> Oh [there is AesGcm in .NET Standard 2.1](https://learn.microsoft.com/en-us/dotnet/api/system.security.cryptography.aesgcm?view=net-9.0#applies-to). Then why not multi-target to .NET Standard 2.1 and forward to the inbox AesGcm?

Because:

> NETStandard 2.1 defines a version of the class with two constructors that we're not exposing from this package (they've since been marked [Obsolete] for "no one uses these in a secure enough manner" reasons). So the problem is that if someone runs on a NS2.1 target that isn't NET8+ they'll get a MissingMethodException, because the underlying runtime won't have exposed the ctors that this library says exist.

Reply:
> Oh [there is AesGcm in .NET Standard 2.1](https://learn.microsoft.com/en-us/dotnet/api/system.security.cryptography.aesgcm?view=net-9.0#applies-to). Then why not multi-target to .NET Standard 2.1 and forward to the inbox AesGcm? Sounds like a no-brainer.

Because the one in .NET Standard 2.1 has constructors that are obsolete, specifically for "people introduce vulnerabilities using them". The only version of `AesGcm` that exists in .NET Standard 2.1 is impossible to use safely, so we don't want to out-box it for .NET Standard 2.1.

Reply:
Can you just leave that type out of the `netstandard` assembly completely?  Why even put it there if you don't have a safe version that can be used everywhere?

Reply:
Because we want it for netstandard2.0, where `AesGcm` does not exist so we can provide the API we want.

Reply:
> The only version of `AesGcm` that exists in .NET Standard 2.1 is impossible to use safely

Ah I see, I now understand the problem.

Reply:
I chatted with @bartonjs offline.  I really don't think it's worth exposing this for netsandard with additional API that will have a known unification problem on netstandard2.1 and .NETCore 3.0-7.0 (even though out of support).

IMO you should either a) leave out the constructors from netstandard or b) leave the type out entirely from netstandard.  You can still provide it on netfx, but those places where it already exists -- you really don't want to pretend it works or has more API than it actually does.